### PR TITLE
FlatGeobuf: Rework/improve error handling

### PR DESCRIPTION
**Pull request by bjornharrtell**
## What does this PR do?

Checks return values for all I/O and memory allocations to make sure to handle potential errors and some tidying up.

### CODE REVIEW

File: gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp [hunk 1]
```diff
@@ -443,9 +443,10 @@ OGRErr OGRFlatGeobufLayer::querySpatialIndex() {
         VSIFree(m_featureOffsets);
         m_featureOffsets = nullptr;
     }
-    auto featureOffsetsSize = static_cast<size_t>(featuresCount) * 8;
+    auto featureOffetsCount = static_cast<size_t>(featuresCount);
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp]: is there a chance that featuresCount could be > std::numeric_limits<size_t> on 32 bit platforms ? (or overflow in the below * 8 multiplication)

File: gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp [hunk 2]
```diff
@@ -443,9 +443,10 @@ OGRErr OGRFlatGeobufLayer::querySpatialIndex() {
...
+    auto featureOffsetsSize = featureOffetsCount * 8;
     m_featureOffsets = static_cast<uint64_t *>(VSI_MALLOC_VERBOSE(featureOffsetsSize));
-    if (VSIFReadL(m_featureOffsets, 8, static_cast<size_t>(featuresCount), m_poFp) != featuresCount)
+    if (VSIFReadL(m_featureOffsets, 8, featureOffetsCount, m_poFp) != featureOffetsCount)
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp]: the above malloc() could potential fail, so null checking should probably be done

File: gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp (hunk 1 above)
Comment (bjornharrtell) [reply gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp]: Hmm yes, both issues are relevant.

The max number of features per spec is 64-bit unsigned but I guess practically limited to to 64-bit unsigned / 8 due the logic you point out on 64-bit arch. I don't see any simple way other than that I make hard checks to require lower number of features than that on read/write. If I do my math correctly it should be 536870912 on 32-bit and 2305843009213690000 on 64-bit so hopefully plenty enough. (Other at this point less known practical limitations notwithstanding)

Comment (rouault) [reply gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp]: I was more thinking about corrupted files than real world ones

Comment (bjornharrtell) [reply gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp]: Ok. I've added checks but only for what the header provide, in this case it's the number of found indices in the spatial index. If the spatial index is corrupted things can surely go very bad, I'll consider if I can add some sanity checks there too.

**Comment by rouault** [comment]
There's a conflict to solve now that I've merged the previous PR

**Comment by bjornharrtell** [comment]
Resolved conflict and rebased.

**Pull request merged**